File: androidApp/src/main/java/dev/mateusz1913/f1results/android/presentation/current_race_results/CurrentRaceResultsViewModel.kt
package dev.mateusz1913.f1results.android.presentation.current_race_results

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.mateusz1913.f1results.datasource.data.qualifying_results.RaceWithQualifyingResultsType
import dev.mateusz1913.f1results.datasource.data.race_results.RaceWithResultsType
import dev.mateusz1913.f1results.datasource.repository.qualifying_results.QualifyingResultsRepository
import dev.mateusz1913.f1results.datasource.repository.race_results.RaceResultsRepository
import io.github.aakira.napier.Napier
import kotlinx.coroutines.launch

class CurrentRaceResultsViewModel(
    private val raceResultsRepository: RaceResultsRepository,
    private val qualifyingResultsRepository: QualifyingResultsRepository,
): ViewModel() {
    var qualifyingResults by mutableStateOf<RaceWithQualifyingResultsType?>(null)
        private set
    var raceResults by mutableStateOf<RaceWithResultsType?>(null)
        private set

    var isFetchingQualifyingResults by mutableStateOf(false)
        private set
    var isFetchingRaceResults by mutableStateOf(false)
        private set

    init {
        fetchCurrentRaceResults()
        fetchCurrentQualifyingResults()
    }

    fun fetchCurrentRaceResults() {
        isFetchingRaceResults = true
        viewModelScope.launch {
            raceResults = raceResultsRepository.fetchRaceResult(season = "current", round = "last")
            isFetchingRaceResults = false
        }
    }

    fun fetchCurrentQualifyingResults() {
        isFetchingQualifyingResults = true
        viewModelScope.launch {
            qualifyingResults = qualifyingResultsRepository.fetchQualifyingResult(season = "current", round = "last")
            isFetchingQualifyingResults = false
        }
    }
}

File: swiftApp/Shared/Presentation/CurrentRaceResults/CurrentRaceResultsViewModel.swift
import SwiftUI
import shared

@MainActor
class CurrentRaceResultsViewModel: ObservableObject {
    @Published var qualifyingResults: RaceWithQualifyingResultsType? = nil
    @Published var raceResults: RaceWithResultsType? = nil
    
    private let qualifyingResultsRepository: QualifyingResultsRepository
    private let raceResultsRepository: RaceResultsRepository
    
    init(raceResultsRepository: RaceResultsRepository, qualifyingResultsRepository: QualifyingResultsRepository) {
        self.raceResultsRepository = raceResultsRepository
        self.qualifyingResultsRepository = qualifyingResultsRepository
        
        fetchCurrentRaceResults()
        fetchCurrentQualificationResults()
    }
    
    func fetchCurrentRaceResults() {
        raceResultsRepository.fetchRaceResult(season: "current", round: "last", position: nil) { response, error in
            if let error = error {
                NapierNapier.shared.d(message: "\(error.localizedDescription)", throwable: nil, tag_: "\(self)")
            } else if let response = response {
                self.raceResults = response
            }
        }
    }
    
    func fetchCurrentQualificationResults() {
        qualifyingResultsRepository.fetchQualifyingResult(season: "current", round: "last", position: nil) { response, error in
            if let error = error {
                NapierNapier.shared.d(message: "\(error.localizedDescription)", throwable: nil, tag_: "\(self)")
            } else if let response = response {
                self.qualifyingResults = response
            }
        }
    }
}
